codalab-service.py

```diff
@@ -0,0 +1,387 @@
+        start_cmd.add_argument('--frontend-port', type=str, help='Port for the React server to listen on (by default it is not exposed to the host machine)', default=argparse.SUPPRESS)
+        start_cmd.add_argument('--mysql-port', type=str, help='Port for the MYSQL database to listen on (by default it is not exposed to the host machine)', default=argparse.SUPPRESS)
+
+        #  SSL CONFIGURATION
+
+        start_cmd.add_argument('--use-ssl', action='store_true', help='If specified set the server up with SSL', default=argparse.SUPPRESS)
+        start_cmd.add_argument('--ssl-cert-file', type=str, help='Path to the cert file for SSL', default=argparse.SUPPRESS)
+        start_cmd.add_argument('--ssl-key-file', type=str, help='Path to the key file for SSL', default=argparse.SUPPRESS)
+
+        #  LOGS SETTINGS
+
+        logs_cmd.add_argument('services', nargs='*', default='all', help='Services to print logs for', choices=['mysql', 'rest-server', 'bundle-manager', 'frontend', 'nginx', 'worker', 'all'])
+        logs_cmd.add_argument('--follow', '-f', action='store_true', help='If specified follow the logs', default=argparse.SUPPRESS)
+        logs_cmd.add_argument('--tail', '-t', type=int, help='If specified tail TAIL lines from the ends of each log', default=argparse.SUPPRESS)
+
+        #  TESTS SETTINGS
+
+        test_cmd.add_argument('tests', metavar='TEST', nargs='+', type=str, choices=test_cli.TestModule.modules.keys() + ['all', 'default'], default=['default'], help='Tests to run')
+
+        #  RUN SETTINGS
+
+        run_cmd.add_argument('service', metavar='SERVICE', type=str, choices=['mysql', 'rest-server', 'bundle-manager', 'frontend', 'nginx', 'worker', 'all'], help='Service container to run command on')
+        run_cmd.add_argument('cmd', metavar='CMD', type=str, help='Command to run')
+        run_cmd.add_argument('--no-root', action='store_true', help='If specified run as current user')
+        return parser
+
+    @classmethod
+    def get_args(cls):
+        parser = cls._get_parser()
+        args = cls()
+        args.apply_environment(os.environ)
+        parser.parse_args(namespace=args)
+        return args
+
+    def __init__(self):
+        for arg in self.DEFAULT_ARGS.keys():
+            setattr(self, arg, None)
+        self.root_dir = os.path.dirname(os.path.realpath(__file__))
+
+    def _apply_defaults(self):
+        for arg, default in self.DEFAULT_ARGS.items():
+            if getattr(self, arg) is None:
+                setattr(self, arg, default)
+
+        if self.worker_dir is None:
+            self.worker_dir = os.path.join(self.root_dir, 'codalab-worker-scratch')
+
+    def apply_environment(self, env):
+        for arg, var in self.ARG_TO_ENV_VAR.items():
+            if var in env:
+                setattr(self, arg, env[var])
+        self._apply_defaults()
+
+
+class CodalabServiceManager(object):
+
+    @staticmethod
+    def resolve_env_vars(args):
+        environment = {
```

Comment:
Document it that environment variables are needed for docker-compose?

Reply:
The users don't need to set env vars anymore, and we don't need to parse env vars actually (in this setup).

Here's some context: 
- The only way to modify  the behavior of docker-compose files is to use environment variables.
- In the earlier version of the script (written in bash), instead of trying to parse many command line arguments and then issue the docker-compose commands with env variables initialized from those, I just passed on the user's current env vars to the command. As a result, the user had to set every important env variable.
- In the python script, the env vars that are applied to the compose commands and the env vars of the user's computer are completely separate. Thanks to the subprocess library, I construct the set of necessary variables for each compose command.
- As a result, it doesn't really matter what env variables are set on the host machine, we can reconstruct all the necessary env variables from the command line arguments of the script before calling the compose command.
- For backwards compatibility, I initially tried to write the python script such that its arguments can be set from environment variables AND command line arguments.
- The precedence there is CLI argument > env var > default.
- What that means is that all arguments have some default which is used if that argument is not set on the command line or the environment. If there's an env var but no CLI argument, the env var is used, if there's an env var and also a command line argument, the command line argument overwrites the value from the env var.
- To do this 3-way merging of arguments (between defaults, env vars and CLI args) I had to do all the custom stuff with mappings (from env var name to arg name) and suppressing the default `default` behavior of argparse (which doesn't play nicely with using a custom Namespace that can have values set at other points).

Our options:
1. Get rid of env var parsing for arguments: We can tell users to just pass in command line arguments for any configuration and get rid of all that messy logic and mappings. I've already rewrote the travis script to use CLI args for testing, we can just move everything to use the CLI args and tell people that's how you configure the codalab instance.

Pros:
    - Get rid of messy argument parsing logic
Cons:
    - If you have some local defaults you want to persist,  there's no way to do that and you need to use a long command with many arguments every time you do something

2. Instead of env vars, use some config file parsing (like JSON).

Pros:
    - Reduce messy merging logic (no more need for mapping env var names to arg names etc)
    - Allow people to save local defaults for easily redeploying etc
Cons:
    - Most of the messy logic still there (since we still need to 3-way merge CLI, JSON and default values)

3. Keep it like this with env vars being the way to persist your defaults

Pros:
    - Local defaults possible
    - No rewriting code portions
Cons:
    - No reduction in mess
    - Env vars are more opaque and may change since last time service started etc. So not great for rerunning etc.


Looking at this, option (2) seems to be the best option, but we can also go with (1) and tell people to save the long commands they use to start the service. 

What do you think @percyliang ?

Reply:
Even if we decide on doing the 3-way merge, I think there should be a way to writing the code so that things are more declarative and defined in one place, so the code looks more like enumerating over arguments/fields/vars rather than having it be flattened.

I think we should definitely support loading from file/environment, not just from command-line args.  On a shared system, the password would show up when any user does a `ps`, which is bad.  I think loading from a config file which is isomorphic to the arg parse arguments would be ideal (you write a simple converter than converts dicts to args).

Reply:
One option here is to use a 3rd party CLI building framework like [click](https://click.palletsprojects.com/en/7.x/) 

The advantage is things like env var based configuration are handled natively and it has builtin handlers for things like passwords, confirmations etc.

The main downside is it is just another library to learn, and in particular click has a decorator-based approach to generating the CLI rather than the more explicit argparse, which would look unfamiliar to any developer used to the argparse style.